runtime/vm/jvminit.c

```diff
@@ -7896,3 +7910,119 @@ parseGlrOption(J9JavaVM* jvm, char* option)
 
 	return JNI_ERR;
 }
+
+static UDATA
+ensureHashedHashFn(void *entry, void *userData)
+{
+	J9UTF8 *hashtableEntry = *(J9UTF8 **)entry;
+
+	return computeHashForUTF8(J9UTF8_DATA(hashtableEntry), J9UTF8_LENGTH(hashtableEntry));
+}
+
+static UDATA
+ensureHashedHashEqualFn(void *lhsEntry, void *rhsEntry, void *userData)
+{
+	J9UTF8 *lhsHashtableEntry = *(J9UTF8 **)lhsEntry;
+	J9UTF8 *rhsHashtableEntry = *(J9UTF8 **)rhsEntry;
+
+	return J9UTF8_DATA_EQUALS(J9UTF8_DATA(lhsHashtableEntry), J9UTF8_LENGTH(lhsHashtableEntry), J9UTF8_DATA(rhsHashtableEntry), J9UTF8_LENGTH(rhsHashtableEntry));
+}
+
+static UDATA
+ensureHashedDoDelete(void *entry, void *userData)
+{
+	PORT_ACCESS_FROM_PORT((J9PortLibrary *)userData);
+
+	j9mem_free_memory(*(J9UTF8 **)entry);
+	return TRUE;
+}
+
+static void
+cleanupEnsureHashedConfig(J9JavaVM* jvm)
+{
+	PORT_ACCESS_FROM_JAVAVM(jvm);
+
+	if (NULL != jvm->ensureHashedClasses) {
+		hashTableForEachDo(jvm->ensureHashedClasses, ensureHashedDoDelete, PORTLIB);
+		hashTableFree(jvm->ensureHashedClasses);
+	}
+	jvm->ensureHashedClasses = NULL;
+}
+
+static UDATA
+parseEnsureHashedOption(J9JavaVM *jvm, char *option)
+{
+	PORT_ACCESS_FROM_JAVAVM(jvm);
+
+	if (0 == strcmp(option, "")) {
+		/* Empty class name */
+		return J9VMDLLMAIN_FAILED;
+	}
+
+	if (NULL == jvm->ensureHashedClasses) {
+		jvm->ensureHashedClasses = hashTableNew(OMRPORT_FROM_J9PORT(PORTLIB),
+				J9_GET_CALLSITE(),
+				16,
+				sizeof(char*),
+				0,
+				0,
+				OMRMEM_CATEGORY_VM,
+				ensureHashedHashFn,
+				ensureHashedHashEqualFn,
+				NULL,
+				PORTLIB);
+
+		if (NULL == jvm->ensureHashedClasses) {
+			/* we failed the allocation */
+			return JNI_ENOMEM;
+		}
+	}
+
+	char *className = option;
+	size_t classNameLength = strlen(className);
+	J9UTF8 *hashtableEntry;
+	J9UTF8 **oldEntry;
+
+	hashtableEntry = j9mem_allocate_memory(classNameLength + sizeof(J9UTF8_LENGTH(hashtableEntry)), OMRMEM_CATEGORY_VM);
+	if (NULL == hashtableEntry) {
+		return JNI_ENOMEM;
+	}
+	memcpy((char *)&J9UTF8_DATA(hashtableEntry)[0], className, classNameLength);
+	J9UTF8_SET_LENGTH(hashtableEntry, (U_16)classNameLength);
+
+	/* now add the new entry if it doesn't exist */
+	oldEntry = hashTableFind(jvm->ensureHashedClasses, &hashtableEntry);
```

Comment:
Do you need to do a find before adding an entry ? I have an impression if an entry already exists in the hashtable, `hashTableAdd()` will just return that one.